### EDA al csv juegos_en_steam

`Steam` es una plataforma de distribución digital desarrollada por la famosa empresa de videojuegos Valve Corporation. Imagina que es como una enorme tienda en línea donde puedes comprar, descargar y jugar una amplia variedad de videojuegos. Desde éxitos populares hasta títulos indie únicos, Steam ofrece algo para todos los gustos.

### Campos del Csv
- appid
    - Aplication Id, Es un numero unico asociado a cada juego en la plataforma Steam
- name
    - Nombre del Video Juego 
- release_date
    - Fecha de Lanzamiento
- english
    - Esta en Ingles?
- developer
    -  Estudio o Empresa que desarrolló el Juego
- publisher
    -  Empresa o entidad que publica y distribuye el juego.
- platforms
    - Sistemas Operativos
- required_age
    - Requiere Mayoria de Edad
- categories
    - Categorias
- genres
    - Genero
- steamspy_tags
    - Etiquetas asociadas al Juego
- achievements
    - Logros, Metas o desafios
- positive_ratings
    - Calificaciones Positivas
- negative_ratings
    - Calificaciones Negativas
- average_playtime
    - Tiempo de Juego Promedio
- median_playtime
    - Mediana del Tiempo de Juego
- owners
    - Cantidad de personas que lo han adquirido
- price
    - Precio del Video Juego






Imoprtando Librerias

In [1]:
import polars as pl
import duckdb

Creo el Dataframe

In [2]:
df = pl.read_csv('csv_originales/03_Juegos_en_steam.csv', separator=',')

In [3]:
df.head()

shape: (5, 18)
┌───────┬─────────────┬─────────────┬─────────┬───┬─────────────┬─────────────┬────────────┬───────┐
│ appid ┆ name        ┆ release_dat ┆ english ┆ … ┆ average_pla ┆ median_play ┆ owners     ┆ price │
│ ---   ┆ ---         ┆ e           ┆ ---     ┆   ┆ ytime       ┆ time        ┆ ---        ┆ ---   │
│ i64   ┆ str         ┆ ---         ┆ i64     ┆   ┆ ---         ┆ ---         ┆ str        ┆ f64   │
│       ┆             ┆ str         ┆         ┆   ┆ i64         ┆ i64         ┆            ┆       │
╞═══════╪═════════════╪═════════════╪═════════╪═══╪═════════════╪═════════════╪════════════╪═══════╡
│ 10    ┆ Counter-Str ┆ 2000-11-01  ┆ 1       ┆ … ┆ 17612       ┆ 317         ┆ 10000000-2 ┆ 7.19  │
│       ┆ ike         ┆             ┆         ┆   ┆             ┆             ┆ 0000000    ┆       │
│ 20    ┆ Team        ┆ 1999-04-01  ┆ 1       ┆ … ┆ 277         ┆ 62          ┆ 5000000-10 ┆ 3.99  │
│       ┆ Fortress    ┆             ┆         ┆   ┆             ┆             ┆ 000000     ┆       │
│       ┆ Classic     ┆             ┆         ┆   ┆             ┆             ┆            ┆       │
│ 30    ┆ Day of      ┆ 2003-05-01  ┆ 1       ┆ … ┆ 187         ┆ 34          ┆ 5000000-10 ┆ 3.99  │
│       ┆ Defeat      ┆             ┆         ┆   ┆             ┆             ┆ 000000     ┆       │
│ 40    ┆ Deathmatch  ┆ 2001-06-01  ┆ 1       ┆ … ┆ 258         ┆ 184         ┆ 5000000-10 ┆ 3.99  │
│       ┆ Classic     ┆             ┆         ┆   ┆             ┆             ┆ 000000     ┆       │
│ 50    ┆ Half-Life:  ┆ 1999-11-01  ┆ 1       ┆ … ┆ 624         ┆ 415         ┆ 5000000-10 ┆ 3.99  │
│       ┆ Opposing    ┆             ┆         ┆   ┆             ┆             ┆ 000000     ┆       │
│       ┆ Force       ┆             ┆         ┆   ┆             ┆             ┆            ┆       │
└───────┴─────────────┴─────────────┴─────────┴───┴─────────────┴─────────────┴────────────┴───────┘

In [4]:
df.shape

(27075, 18)

Examinando los Valores Nulos

In [5]:
for column in df.columns:
    print(f"Columna: {column} Cantidad de nulos -->  {df[column].is_null().sum()}")

Columna: appid Cantidad de nulos -->  0
Columna: name Cantidad de nulos -->  0
Columna: release_date Cantidad de nulos -->  0
Columna: english Cantidad de nulos -->  0
Columna: developer Cantidad de nulos -->  0
Columna: publisher Cantidad de nulos -->  0
Columna: platforms Cantidad de nulos -->  0
Columna: required_age Cantidad de nulos -->  0
Columna: categories Cantidad de nulos -->  0
Columna: genres Cantidad de nulos -->  0
Columna: steamspy_tags Cantidad de nulos -->  0
Columna: achievements Cantidad de nulos -->  0
Columna: positive_ratings Cantidad de nulos -->  0
Columna: negative_ratings Cantidad de nulos -->  0
Columna: average_playtime Cantidad de nulos -->  0
Columna: median_playtime Cantidad de nulos -->  0
Columna: owners Cantidad de nulos -->  0
Columna: price Cantidad de nulos -->  0


In [ ]:
df.head(1)

Averiguando cuantos Generos existen

In [6]:
print(df['genres'].unique().count())
print(df['categories'].unique().count())

1552
3333


En realidad no existen 1552 Generos, simplemente hay registradas 1552 combinaciones de Generos, ya que hay video juegos que pertenecen a mas de un Genero, por lo que el '.unique()` lo ve como otro genero nuevo. Por ejemplo hay Juegos que pertenecen a Categorias de 'Accion', 'Primera Persona', 'Disparo', etc, El Mismo análisis tambien aplica para las Categorias

***
Spliteo y Extraigo las Distintas Categorias, Generos y Plataformas para poder saber de cuantos estamos hablando

In [7]:
categories_list = []
genres_list = []
platforms_list = []
for i in range(len(df)):
    categories = df[i,8].split(";")
    genres = df[i,9].split(";")
    platforms = df[i,6].split(";")
    for category in categories:
        if not category in categories_list:
            categories_list.append(category)
    for gen in genres:
        if not gen in genres_list:
            genres_list.append(gen)
    for platform in platforms:
        if not platform in platforms_list:
            platforms_list.append(platform) 

print(f"Numero de Categorias -->  {len(categories_list)}")
print(f"Numero de Generos -->  {len(genres_list)}")
print(f"Numero de Plataformas--> {len(platforms_list)}")

Numero de Categorias -->  29
Numero de Generos -->  29
Numero de Plataformas--> 3


In [8]:
print(genres_list)
print(categories_list)
print(platforms_list)
#Las siglas “RPG” provienen del inglés y significan “Role-Playing Game”, que en español se traduce como “juego de rol”.

['Action', 'Free to Play', 'Strategy', 'Adventure', 'Indie', 'RPG', 'Animation & Modeling', 'Video Production', 'Casual', 'Simulation', 'Racing', 'Violent', 'Massively Multiplayer', 'Nudity', 'Sports', 'Early Access', 'Gore', 'Utilities', 'Design & Illustration', 'Web Publishing', 'Education', 'Software Training', 'Sexual Content', 'Audio Production', 'Game Development', 'Photo Editing', 'Accounting', 'Documentary', 'Tutorial']
['Multi-player', 'Online Multi-Player', 'Local Multi-Player', 'Valve Anti-Cheat enabled', 'Single-player', 'Steam Cloud', 'Steam Achievements', 'Steam Trading Cards', 'Captions available', 'Partial Controller Support', 'Includes Source SDK', 'Cross-Platform Multiplayer', 'Stats', 'Commentary available', 'Includes level editor', 'Steam Workshop', 'In-App Purchases', 'Co-op', 'Full controller support', 'Steam Leaderboards', 'SteamVR Collectibles', 'Online Co-op', 'Shared/Split Screen', 'Local Co-op', 'MMO', 'VR Support', 'Mods', 'Mods (require HL2)', 'Steam Turn N

***
Creare una Matrices para Generos, Categorias y Plataformas. Por Ejemplo la matriz de Generos contara con 32 columnas `appid`, `possitive_ratings`,`negative_ratings` y `TODOS LOS GENEROS POR SEPARADO` colocando un 1 solo solo en los generos a los cuales pertenece cada Genero

In [9]:
df[0]

shape: (1, 18)
┌───────┬─────────────┬─────────────┬─────────┬───┬─────────────┬─────────────┬────────────┬───────┐
│ appid ┆ name        ┆ release_dat ┆ english ┆ … ┆ average_pla ┆ median_play ┆ owners     ┆ price │
│ ---   ┆ ---         ┆ e           ┆ ---     ┆   ┆ ytime       ┆ time        ┆ ---        ┆ ---   │
│ i64   ┆ str         ┆ ---         ┆ i64     ┆   ┆ ---         ┆ ---         ┆ str        ┆ f64   │
│       ┆             ┆ str         ┆         ┆   ┆ i64         ┆ i64         ┆            ┆       │
╞═══════╪═════════════╪═════════════╪═════════╪═══╪═════════════╪═════════════╪════════════╪═══════╡
│ 10    ┆ Counter-Str ┆ 2000-11-01  ┆ 1       ┆ … ┆ 17612       ┆ 317         ┆ 10000000-2 ┆ 7.19  │
│       ┆ ike         ┆             ┆         ┆   ┆             ┆             ┆ 0000000    ┆       │
└───────┴─────────────┴─────────────┴─────────┴───┴─────────────┴─────────────┴────────────┴───────┘

In [10]:
data_generos = []
data_categorias = []
data_plataformas = []
data_titulos = []
for i in range(len(df)):
    #Extrayendo o Haciendo un Spliteo para obtener Data de Generos Individualizados
    diccionario_genero = {}
    generos = df[i,9].split(";")  
    diccionario_genero['appid'] = df[i,0]
    diccionario_genero['possitive_ratings'] = df[i,12]
    diccionario_genero['negative_ratings'] = df[i,13]
    for genero in generos:
        diccionario_genero[genero] = 1
    data_generos.append(diccionario_genero) 

    #Extrayendo o Haciendo un Spliteo para obtener Data de Categorias Individualizadas
    diccionario_categorias = {}
    categorias = df[i,8].split(";")
    diccionario_categorias['appid'] = df[i,0]
    diccionario_categorias['possitive_ratings'] = df[i,12]
    diccionario_categorias['negative_ratings'] = df[i,13]
    for categoria in categorias:
        diccionario_categorias[categoria] = 1
    data_categorias.append(diccionario_categorias)

    #Extrayendo o Haciendo un Spliteo para obtener Data de Plataformas Individualizadas
    diccionario_plataformas = {}
    plataformas = df[i,6].split(";")
    diccionario_plataformas['appid'] = df[i,0]
    for plataforma in plataformas:
        diccionario_plataformas[plataforma] = 1
    data_plataformas.append(diccionario_plataformas)    



En este caso se Observa que el Video Juego con el Appid 3560 tiene 98 calificaciones positivas, 18 calificaciones negativas y pertenece a 2 Generos, 'Estrategia' y 'Casual'

In [11]:
data_generos[100]

{'appid': 3560,
 'possitive_ratings': 98,
 'negative_ratings': 18,
 'Strategy': 1,
 'Casual': 1}

In [12]:
data_categorias[1]

{'appid': 20,
 'possitive_ratings': 3318,
 'negative_ratings': 633,
 'Multi-player': 1,
 'Online Multi-Player': 1,
 'Local Multi-Player': 1,
 'Valve Anti-Cheat enabled': 1}

In [13]:
data_plataformas[0]

{'appid': 10, 'windows': 1, 'mac': 1, 'linux': 1}

Creo el Schema de la Matriz Individualizada de Generos

In [14]:

schema_genero = {'appid': pl.Int64, 'possitive_ratings':pl.Int64, 'negative_ratings':pl.Int64}
for element in genres_list:
    schema_genero[element] = pl.Int64

schema_categoria = {'appid': pl.Int64, 'possitive_ratings':pl.Int64, 'negative_ratings':pl.Int64}
for element in categories_list:
    schema_categoria[element] = pl.Int64

print(len(schema_categoria))
print(len(schema_genero))

32
32


Creo los dataframes de Generos (Matriz)

In [15]:
df_generos = pl.DataFrame(data=data_generos, nan_to_null=True, strict=False, schema=schema_genero)

In [16]:
df_categorias = pl.DataFrame(data=data_categorias, nan_to_null=True, strict=False, schema=schema_categoria)

In [17]:
df_plataformas = pl.DataFrame(data=data_plataformas, nan_to_null=True, strict=False)

Obviamente donde Los Generos, Categorias o Plataformas No coincidan con los Video Juegos hay un null que debe ser reemplazado por un 0

In [18]:
df_generos = df_generos.fill_null(0)

In [19]:
df_categorias = df_categorias.fill_null(0)

In [20]:
df_plataformas = df_plataformas.fill_null(0)

Ya tenemos las Matrices de Generos, Categorias y Plataformas Donde Podemos Hacer una ponderacion Individual de los ratings por Genero

In [21]:
df_generos[:-1]

shape: (27_074, 32)
┌─────────┬─────────────┬─────────────┬────────┬───┬─────────┬────────────┬─────────────┬──────────┐
│ appid   ┆ possitive_r ┆ negative_ra ┆ Action ┆ … ┆ Photo   ┆ Accounting ┆ Documentary ┆ Tutorial │
│ ---     ┆ atings      ┆ tings       ┆ ---    ┆   ┆ Editing ┆ ---        ┆ ---         ┆ ---      │
│ i64     ┆ ---         ┆ ---         ┆ i64    ┆   ┆ ---     ┆ i64        ┆ i64         ┆ i64      │
│         ┆ i64         ┆ i64         ┆        ┆   ┆ i64     ┆            ┆             ┆          │
╞═════════╪═════════════╪═════════════╪════════╪═══╪═════════╪════════════╪═════════════╪══════════╡
│ 10      ┆ 124534      ┆ 3339        ┆ 1      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 20      ┆ 3318        ┆ 633         ┆ 1      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 30      ┆ 3416        ┆ 398         ┆ 1      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 40      ┆ 1273        ┆ 267         ┆ 1      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 50      ┆ 5250        ┆ 288         ┆ 1      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ …       ┆ …           ┆ …           ┆ …      ┆ … ┆ …       ┆ …          ┆ …           ┆ …        │
│ 1065160 ┆ 2           ┆ 2           ┆ 0      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 1065230 ┆ 3           ┆ 0           ┆ 0      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 1065570 ┆ 8           ┆ 1           ┆ 1      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 1065650 ┆ 0           ┆ 1           ┆ 1      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
│ 1066700 ┆ 2           ┆ 0           ┆ 0      ┆ … ┆ 0       ┆ 0          ┆ 0           ┆ 0        │
└─────────┴─────────────┴─────────────┴────────┴───┴─────────┴────────────┴─────────────┴──────────┘

In [22]:
df_categorias[:-1]

shape: (27_074, 32)
┌─────────┬────────────┬────────────┬────────────┬───┬────────────┬──────┬────────────┬────────────┐
│ appid   ┆ possitive_ ┆ negative_r ┆ Multi-play ┆ … ┆ VR Support ┆ Mods ┆ Mods       ┆ Steam Turn │
│ ---     ┆ ratings    ┆ atings     ┆ er         ┆   ┆ ---        ┆ ---  ┆ (require   ┆ Notificati │
│ i64     ┆ ---        ┆ ---        ┆ ---        ┆   ┆ i64        ┆ i64  ┆ HL2)       ┆ ons        │
│         ┆ i64        ┆ i64        ┆ i64        ┆   ┆            ┆      ┆ ---        ┆ ---        │
│         ┆            ┆            ┆            ┆   ┆            ┆      ┆ i64        ┆ i64        │
╞═════════╪════════════╪════════════╪════════════╪═══╪════════════╪══════╪════════════╪════════════╡
│ 10      ┆ 124534     ┆ 3339       ┆ 1          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 20      ┆ 3318       ┆ 633        ┆ 1          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 30      ┆ 3416       ┆ 398        ┆ 1          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 40      ┆ 1273       ┆ 267        ┆ 1          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 50      ┆ 5250       ┆ 288        ┆ 1          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ …       ┆ …          ┆ …          ┆ …          ┆ … ┆ …          ┆ …    ┆ …          ┆ …          │
│ 1065160 ┆ 2          ┆ 2          ┆ 0          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 1065230 ┆ 3          ┆ 0          ┆ 0          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 1065570 ┆ 8          ┆ 1          ┆ 0          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 1065650 ┆ 0          ┆ 1          ┆ 1          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
│ 1066700 ┆ 2          ┆ 0          ┆ 0          ┆ … ┆ 0          ┆ 0    ┆ 0          ┆ 0          │
└─────────┴────────────┴────────────┴────────────┴───┴────────────┴──────┴────────────┴────────────┘

In [23]:
df_plataformas

shape: (27_075, 4)
┌─────────┬─────────┬─────┬───────┐
│ appid   ┆ windows ┆ mac ┆ linux │
│ ---     ┆ ---     ┆ --- ┆ ---   │
│ i64     ┆ i64     ┆ i64 ┆ i64   │
╞═════════╪═════════╪═════╪═══════╡
│ 10      ┆ 1       ┆ 1   ┆ 1     │
│ 20      ┆ 1       ┆ 1   ┆ 1     │
│ 30      ┆ 1       ┆ 1   ┆ 1     │
│ 40      ┆ 1       ┆ 1   ┆ 1     │
│ 50      ┆ 1       ┆ 1   ┆ 1     │
│ …       ┆ …       ┆ …   ┆ …     │
│ 1065230 ┆ 1       ┆ 0   ┆ 0     │
│ 1065570 ┆ 1       ┆ 0   ┆ 0     │
│ 1065650 ┆ 1       ┆ 0   ┆ 0     │
│ 1066700 ┆ 1       ┆ 1   ┆ 0     │
│ 1069460 ┆ 1       ┆ 1   ┆ 0     │
└─────────┴─────────┴─────┴───────┘

In [ ]:
# Verificando como voy a obtener los nombres de las columnas dentro de los ciclos for
print(df_generos[:,3].name)
print(df_categorias[:,3].name)

In [ ]:
df_categorias

***
Aca, puedo sencillamente crear la Metrica que crea Conveniente para caso de Estudio
* Tablas Generos y Categorias
    * diferencial_ratings : Diferencia entre calificaciones Positivas y Negativas
    * percent_possitive: Porcentaje de Calificaciones Positivas con respecto al total

In [24]:
df_generos = df_generos.with_columns(
    diferencial_ratings = pl.col('possitive_ratings') - pl.col('negative_ratings')
)

df_generos = df_generos.with_columns(
    percent_possitive = pl.col('possitive_ratings') / (pl.col('possitive_ratings') + pl.col('negative_ratings'))
)

In [25]:
df_categorias = df_categorias.with_columns(
    diferencial_ratings = pl.col('possitive_ratings') - pl.col('negative_ratings')
)

df_categorias = df_categorias.with_columns(
    percent_possitive = pl.col('possitive_ratings') / (pl.col('possitive_ratings') + pl.col('negative_ratings'))
)

In [26]:
df = df.drop('platforms').drop('categories').drop('genres')

Hasta aca lo que se hizo fue Splitear a Generos, Categorias y Plataformas para poder estudiarlos por separado
de una tabla (o archivo csv) se convienrten en 4 donde permanecen enlazados por el `id del videojuego appid`

In [27]:
df.write_csv('data/03_steam_videojuegos.csv')
df_categorias.write_csv('data/03_steam_categorias_videojuegos.csv')
df_generos.write_csv('data/03_steam_generos_videojuegos.csv')
df_plataformas.write_csv('data/03_steam_plataformas_videojuegos.csv')

In [28]:
df_generos.shape

(27075, 34)